modules/mono/glue/GodotSharp/GodotSharp/Core/Bridge/ScriptManagerBridge.cs

```diff
@@ -258,9 +258,9 @@ internal static void SetGodotObjectPtr(IntPtr gcHandlePtr, IntPtr newPtr)
                 }
             }
 
-            static Type? GetTypeByGodotClassAttr(Assembly assembly, string nativeTypeNameStr)
+            static Type? GetTypeByGodotClassAttr(Assembly? assembly, string nativeTypeNameStr)
```

Comment:
I've reverted this regardless, but I'm not sure I follow. Couldn't `editorAssembly` only realistically not be null if either `FirstOrDefault()` is given a proper default value or, if finding the assembly is a guarantee, it be swapped out for `First()` instead? It seems like otherwise it's implicitly a nullable reference type
```

Reply:
https://github.com/godotengine/godot/blob/60f3b7967cbd00b4e1f52d33d372646f7bec02f6/modules/mono/glue/GodotSharp/GodotSharp/Core/Bridge/ScriptManagerBridge.cs#L251-L252

`editorAssembly` is null if the `GodotSharpEditor` assembly can't be found. `FirstOrDefault` will return the first value of the enumeration or, if the enumeration is empty, the default value for the type (which in this case is `Assembly`). The default value for `Assembly` (and for every reference type) is `null`.

In short, `editorAssembly` will be null if the enumeration is empty, that is, there is no loaded assembly for which the condition resolves to true, so there's no `GodotSharpEditor` assembly. This can happen when we're executing an exported game.

https://github.com/godotengine/godot/blob/60f3b7967cbd00b4e1f52d33d372646f7bec02f6/modules/mono/glue/GodotSharp/GodotSharp/Core/Bridge/ScriptManagerBridge.cs#L253

This line gets the type `Godot.{nativeTypeNameStr}` from the `editorAssembly` but only if the assembly is not null because we are using the [`?.` operator](https://learn.microsoft.com/en-us/dotnet/csharp/language-reference/operators/conditional-operator).

This means that the value of `wrapperType` can be one of:
- The type we are looking for, if `editorAssembly` is not null and the type was found.
- `null`, if `editorAssembly` is not null and the type was not found.
- `null`, if `editorAssembly` is null.

So in the next line:

https://github.com/godotengine/godot/blob/60f3b7967cbd00b4e1f52d33d372646f7bec02f6/modules/mono/glue/GodotSharp/GodotSharp/Core/Bridge/ScriptManagerBridge.cs#L255-L258

If we enter the if block, that means `wrapperType` is not null but `wrapperType` can only ever be non-null if `editorAssembly` is also not null. We know this because, as explained earlier, if `editorAssembly` was null we could not have found the type we were looking for.

However, the null flow analysis doesn't know this because it doesn't track this far. Null flow analysis will only now that `editorAssembly` is not null if you check directly. See: https://github.com/dotnet/roslyn/issues/48354.

Reply:
I must be missing something crazy obvious, because this makes perfect sense up until this part

https://github.com/godotengine/godot/blob/60f3b7967cbd00b4e1f52d33d372646f7bec02f6/modules/mono/glue/GodotSharp/GodotSharp/Core/Bridge/ScriptManagerBridge.cs#L255

Isn't this checking if `wrapperType` **IS** null? How can we be entering this block with a non-null value? Wouldn't this have to be `wrapperType != null` instead?

Reply:
No, you are right. My bad. I totally misread it.